pkg/backend/openshiftcluster.go

```diff
@@ -203,3 +225,31 @@ func (ocb *openShiftClusterBackend) endLease(stop func(), doc *api.OpenShiftClus
 	_, err = ocb.db.OpenShiftClusters.EndLease(doc.Key, provisioningState, failedProvisioningState)
 	return err
 }
+
+// handleDelete handles document state during context cancelation and backend
+// worker shutdown. If backend worker context is canceled, it will mark document
+// as Failed. We need to wait for it and update to Deleting.
+// End to end state change flow should be:
+// Creating -> Deleting -> Failed -> Deleting
+func (ocb *openShiftClusterBackend) handleDelete(cancel context.CancelFunc, doc *api.OpenShiftClusterDocument) error {
+	cancel()
+	return wait.PollImmediate(time.Second, time.Minute*5, func() (bool, error) {
```

Comment:
yuck.  I think you should be doing something like:
* heartbeat realises deleting has been set
* heartbeat sends cancel() but keeps beating
* worker exits, stops the heartbeat(), clears the lease, doesn't set failed
...
* document is newly dequeued and deletion starts